### PR TITLE
Remove legacy constructor calls from pytorch codebase.

### PR DESCRIPTION
**Pull request by ysiraichi**
Follow up from #53889 
Related to #47112

Removing every occurrence of the legacy constructor call present in PyTorch at:
- _docs_
- _benchmarks_
- _test_
- _caffe2_
- _CONTRIBUTING.md_

### CODE REVIEW

File: benchmarks/overrides_benchmark/common.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ def __init__(self, data, requires_grad=False):
             self._tensor = data
             return
 
-        self._tensor = torch.Tensor(data, requires_grad)
+        self._tensor = torch.empty(data, requires_grad)
```
Comment (rgommers) [inline benchmarks/overrides_benchmark/common.py:17]: This doesn't look right, should be `torch.tensor`?

File: caffe2/python/operator_test/heatmap_max_keypoint_op_test.py [hunk 2]
```diff
@@ -30,8 +30,8 @@ def heatmap_approx_keypoint_ref(maps, rois):
 
 def c10_op_ref(maps, rois):
     keypoints = torch.ops._caffe2.HeatmapMaxKeypoint(
-        torch.Tensor(maps),
-        torch.Tensor(rois),
+        torch.tensor(maps),
+        torch.tensor(rois),
```
Comment (rgommers) [inline caffe2/python/operator_test/heatmap_max_keypoint_op_test.py:34]: The main question I have for reviewing these `Tensor -> tensor` changes is: the dtype of the output can change from `float32` (the default for `Tensor` with e.g. a list of integers) to `int64`. It's not straightforward to see whether that matters or not. Did you check whether `HeatmapMaxKeypoint` behaves the same, or that all callers of `c10_op_ref` don't feed in integers here?

**Review by rgommers — COMMENTED** [review]
Thanks @ysiraichi. Overall looks good; my `Tensor -> tensor` default dtype question applies to all similar changes - have you verified whether the different behavior matters?

File: benchmarks/overrides_benchmark/common.py (hunk 1 above)
Comment (rgommers) [reply benchmarks/overrides_benchmark/common.py:17]: There's lots of similar changes in gh-53889. What am I missing?

Comment (ysiraichi) [reply benchmarks/overrides_benchmark/common.py:17]: AFAIK, this one in particular is called from either two places:

https://github.com/pytorch/pytorch/blob/f2b4b0e9eb670ad79f4460c0d45174f879b523f9/benchmarks/overrides_benchmark/pyspybench.py#L24-L25

https://github.com/pytorch/pytorch/blob/f2b4b0e9eb670ad79f4460c0d45174f879b523f9/benchmarks/overrides_benchmark/bench.py#L54-L55

The behavior of `torch.Tensor` with a scalar is the same as that of `torch.empty`. `torch.tensor` would make sense if `data` was something iterable.

File: caffe2/python/operator_test/heatmap_max_keypoint_op_test.py (hunk 2 above)
Comment (ysiraichi) [reply caffe2/python/operator_test/heatmap_max_keypoint_op_test.py:34]: Yes, I verified the callers of `c10_op_ref` (which, apparently is only one):

https://github.com/pytorch/pytorch/blob/f2b4b0e9eb670ad79f4460c0d45174f879b523f9/caffe2/python/operator_test/heatmap_max_keypoint_op_test.py#L136-L141

Which calls this function with an expanded version of the `input`:

https://github.com/pytorch/pytorch/blob/f2b4b0e9eb670ad79f4460c0d45174f879b523f9/caffe2/python/hypothesis_test_util.py#L647

File: benchmarks/overrides_benchmark/common.py (hunk 1 above)
Comment (ysiraichi) [reply benchmarks/overrides_benchmark/common.py:17]: That said, maybe the intended behavior was `torch.Tensor([1])` and `torch.Tensor([2])`, which would make `torch.tensor` the correct one.

File: caffe2/python/operator_test/layer_norm_op_test.py [hunk 3]
```diff
@@ -344,7 +344,7 @@ def jit_layer_norm(
             expected_norm, expected_mean, expected_std = \
                 _layer_norm_with_affine_ref(axis, eps, X, gamma, beta)
             actual_norm, actual_mean, actual_std = jit_layer_norm(
-                torch.Tensor(X), torch.tensor(gamma), torch.tensor(beta),
+                torch.tensor(X), torch.tensor(gamma), torch.tensor(beta),
```
Comment (ysiraichi) [inline caffe2/python/operator_test/layer_norm_op_test.py:347]: Given that `X` is also a tensor, maybe it would be better to call `X.clone().detach()` (as the warning suggests). For consistency with the else case, I left it as `torch.tensor(X)`, though. Any thoughts?

File: test/test_fx.py [hunk 4]
```diff
@@ -468,7 +468,7 @@ def lower_to_elementwise_interpreter(orig_mod : torch.nn.Module) -> torch.nn.Mod
                             # Pull out constants. These constants will later be
                             # fed to the interpreter C++ object via add_constant()
                             arg_name = f'constant_{constant_idx}'
-                            constants[arg_name] = torch.Tensor(
+                            constants[arg_name] = torch.tensor(
```
Comment (ysiraichi) [inline test/test_fx.py:471]: Since we are pulling out constants, I guess we don't have to worry about their types, right?

File: test/test_jit.py [hunk 5]
```diff
@@ -11209,8 +11209,8 @@ def test_remove_dropout(self):
         class M(torch.nn.Module):
             def __init__(self):
                 super(M, self).__init__()
-                self.weight_0 = torch.nn.Parameter(torch.Tensor(torch.rand(weight_0_shape)))
-                self.weight_1 = torch.nn.Parameter(torch.Tensor(torch.rand(weight_1_shape)))
+                self.weight_0 = torch.nn.Parameter(torch.rand(weight_0_shape))
```
Comment (ysiraichi) [inline test/test_jit.py:11212]: There are lots of things similar to this one:

```python
... = torch.nn.Parameter(torch.Tensor(torch.rand(weight_0_shape)))
```

It seems redundant to me, so I removed the `Tensor` call. Is there a reason for wrapping `rand` into a `Tensor`?

File: caffe2/python/operator_test/layer_norm_op_test.py (hunk 3 above)
Comment (ysiraichi) [reply caffe2/python/operator_test/layer_norm_op_test.py:347]: Actually, `torch.Tensor(X)` creates an alias to `X` (it does not clones its contents as `torch.tensor(X)`). But, looking at the _else_ part, it makes me believe that cloning was the intended behavior...

Comment (rgommers) [reply caffe2/python/operator_test/layer_norm_op_test.py:347]: +1 for using `clone().detach()`

Comment (rgommers) [reply caffe2/python/operator_test/layer_norm_op_test.py:347]: you never want to leave in code or docs that generate warnings

File: test/test_jit.py (hunk 5 above)
Comment (rgommers) [reply test/test_jit.py:11212]: There's no reason, that just looks very weird. Looks like a good cleanup to me.

File: benchmarks/overrides_benchmark/common.py (hunk 1 above)
Comment (rgommers) [reply benchmarks/overrides_benchmark/common.py:17]: I'd ensure the input is a list (or tensor/tensorlike).

The second file you link to has another use of the legacy constructor that needs to be removed:
```
    types = torch.Tensor, SubTensor, WithTorchFunction, SubWithTorchFunction

    for t in types:
        tensor_1 = t(1)
        tensor_2 = t(2)
```

File: test/test_fx.py (hunk 4 above)
Comment (rgommers) [reply test/test_fx.py:471]: this looks fine to me

File: caffe2/python/operator_test/layer_norm_op_test.py (hunk 3 above)
Comment (ysiraichi) [reply caffe2/python/operator_test/layer_norm_op_test.py:347]: I've just realized that `X` there is not a `Tensor`. The `X: torch.Tensor` we see above is a nested function: `jit_layer_norm`. Inside the actual function `test_layer_norm_op_jit`, `X` is a numpy array. So, I guess, `torch.tensor(X)` is the way to go.

File: benchmarks/overrides_benchmark/common.py (hunk 1 above)
Comment (ysiraichi) [reply benchmarks/overrides_benchmark/common.py]: That makes sense.
By the way, what should happen with `torch.Tensor` sub-classes, such as `SubTensor` and `SubWithTorchFunction`, then? They will call the legacy constructor, right!?

Comment (rgommers) [reply benchmarks/overrides_benchmark/common.py]: That's a great question. Yes, they will call the legacy constructor, and it'll be a little hard to avoid. 

https://pytorch.org/docs/stable/notes/extending.html explains how to write a basic subclass - with the deprecated constructor that's no longer the best option. I think the right solution is make the subclass have a `__new__` method, which uses `torch.tensor` to instantiate a tensor, and then create the subclass via taking a custom view of it (similar to what `numpy.ma.MaskedArray` does, see https://github.com/numpy/numpy/blob/main/numpy/ma/core.py#L2836). Until we sort that out, subclasses have to live with the deprecation warning.

I suggest opening a separate issue for that.

**Comment by rgommers** [comment]
The one CI failure on ROCm is unrelated, this is ready to land. @mruberry could you please have a look at this PR?

**Review by mruberry — APPROVED** [review]
Impressively thorough, @ysiraichi, this is great!

**Comment by mruberry** [comment]
What's your thinking for a next step, @ysiraichi?

**Comment by ysiraichi** [comment]
@mruberry So, I guess deprecating the legacy constructor caused creating instances of subclasses of `Tensor` to raise the warning introduced in #54414. I opened an issue #55780 for addressing it.

**Comment by heitorschueroff** [comment]
> @mruberry So, I guess deprecating the legacy constructor caused creating instances of subclasses of `Tensor` to raise the warning introduced in #54414 (which I believe are not desirable). I opened an issue #55780 for addressing it.

@ysiraichi Thank you for opening this issue. Indeed we do not want to raise warning for supported features. This was my fault for not thoroughly considering all the use cases of the Tensor constructor. I appreciate your effort in helping make PyTorch better but for now I am going to unland the PR that added this warning (https://github.com/pytorch/pytorch/pull/54414). We should follow up on another way to deprecate this without affecting valid use cases.
